pkg/otelutil/log/debounce.go

```diff
@@ -0,0 +1,42 @@
+package log
+
+import (
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+)
+
+type Debouncer struct {
+	lastLines map[string]time.Time
+	period    time.Duration
+	mutex     sync.Mutex
+}
+
+func NewDebouncer(period time.Duration) *Debouncer {
+	return &Debouncer{
+		lastLines: make(map[string]time.Time),
+		period:    period,
+	}
+}
+
+func (d *Debouncer) Error(logger logr.Logger, err error, msg string, keysAndValues ...any) {
+	if err == nil {
+		logger.Error(err, msg, keysAndValues...)
+		return
+	}
+
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	if time.Since(d.lastLines[err.Error()]) > d.period {
```

Comment:
Actually can we either but a bound on this or use `msg` as the key instead? As we're using structured logging, it's much less likely that `msg` will change over time. `err` I'm less certain about.

Reply:
The place where this is plumbed in right now have the same msg but two different errors for different types. The error messages stay constant. This is primarily just so we don’t flood logs when people don’t install optional CRDs like ServiceMonitors. Running some simple tests take the log span down from hundreds of entries to like 4.